Playwright: add mobile support for interacting with WPCOM sidebar.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR implements ability for mobile viewports to interact with the WPCOM sidebar.

In previous iterations of the `SidebarComponent`, it assumed the sidebar was visible at all times - which is the case for `desktop` viewports, but not for `mobile`.

- add new methods to click the `My Sites` text on top left of WPCOM dashboard to toggle into view the sidebar.
- fix a critical logic error in `SidebarComponent.gotoMenu` that caused the selectors to be overwritten if both `item` and `subitem` were supplied.

#### Testing instructions

TeamCity
- ensure all tests pass.

worldomonation committed:
Skeleton of the mobile support for SidebarComponent.
- add codepath in the `_click` method to check if mobile, then take additional actions to show the sidebar.

NavbarComponent
- add new method to click on the my site text.
- refactor the click method into its own function shared between all the `click*` methods

worldomonation committed:
Rebase and addresse ESLint issues.

worldomonation committed:
Fix a critical logic error with SidebarComponent.gotoMenu.
- if both `item` and `subitem` was supplied the selector for `subitem` was overwriting the selector for `item`, resulting in Playwright looking for the selector for `subitem` only.

WunderBart commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -11,6 +11,7 @@ import { Page } from 'playwright';
 const selectors = {
 	navbar: '.masterbar',
 	newPostButton: '.masterbar__item-new',
+	mySite: 'text=My Site',
```
```suggestion
	mySitesButton: 'text=My Site',
```

worldomonation replied:
Done.

WunderBart commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -27,13 +28,32 @@ export class NavbarComponent extends BaseContainer {
+	async _click( selector: string ): Promise< void > {
+		const elementHandle = await this.page.waitForSelector( selector );
+		await elementHandle.waitForElementState( 'stable' );
+		await elementHandle.click();
```
Why do we need to wait for the element to be stable? It looks like it's not animated so this check should not be necessary here

worldomonation replied:
I felt it is good practice to explicitly check for element stability. In this case though, I tested locally and found it did not make a difference, so I have removed it.

WunderBart replied:
> I tested locally and found it did not make a difference, so I have removed it.

Nice 👍 It seems like we don't need this `_click` helper anymore then? We can directly call
```js
await this.page.click( selector );`
```
which is equivalent to
```js
const elementHandle = await this.page.waitForSelector( selector );
await elementHandle.click();
```

worldomonation replied:
I think that's fair, let's try it. I'm surprised the navbar component isn't kicking up a fuss about this though, given the troubles I had with it when I was initially developing the new post flow.

WunderBart commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -27,13 +28,32 @@ export class NavbarComponent extends BaseContainer {
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async clickMySite(): Promise< void > {
```
```suggestion
	async clickMySites(): Promise< void > {
```

worldomonation replied:
~~I am not sure if this is a good idea.~~

~~The selector suggests 'site**s**' but the user facing attribute is that of singular 'site'.~~

Hmm, I thought it was singular, but it's plural. Sites it is!

worldomonation replied:
Done!

WunderBart commented on packages/calypso-e2e/src/lib/components/navbar-component.ts:
```diff
@@ -27,13 +28,32 @@ export class NavbarComponent extends BaseContainer {
+	}
+
+	/**
+	 * Clicks on `My Sites`.
```
I think that it's not necessary to always add the doc description. The function describes itself in this case pretty well!

worldomonation replied:
There is a ESLint rule added that requires JSDoc on all functions, so this is to satisfy that.

WunderBart replied:
Does it require the description part as well or just the `@` fields?

worldomonation replied:
Good question, I ran a quick test locally and it _seems_ to be placated by just having documentation of the @ fields. 

As with any rules people will find a way around having to do it, but I am hoping that at least requiring _some_ documentation will make things easier to understand for folks coming in later.

WunderBart replied:
Ah, I meant this as a good thing, actually! I don't think we always need the description part, i.e. we can safely skip it for [the `constructor` method](https://github.com/Automattic/wp-calypso/blob/3d16589b79ee7d48d37ad4c0ac9bbb2e20613eb6/packages/calypso-e2e/src/lib/components/comments-component.ts#L31-L33). 👍 No need to fix anything here though.

WunderBart commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -63,19 +74,26 @@ export class SidebarComponent extends BaseContainer {
 	 */
 	async gotoMenu( { item, subitem }: { item?: string; subitem?: string } ): Promise< void > {
 		let selector;
+		const viewport_name = getViewportName();
```
Why the underscore case here?

worldomonation replied:
That's a good question, was using the environment variable plus reverted to my old Python ways as I was a bit exhausted when writing this.

Will be corrected in next push.

WunderBart commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -86,15 +104,24 @@ export class SidebarComponent extends BaseContainer {
-
-		await this._click( selector );
+	/**
+	 * Opens the sidebar into view.
```
Would be good to note that it's true for mobile viewport only

worldomonation replied:
Done.

WunderBart commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -86,15 +104,24 @@ export class SidebarComponent extends BaseContainer {
+	async _openSidebar(): Promise< void > {
+		const navbarComponent = await NavbarComponent.Expect( this.page );
+		await navbarComponent.clickMySite();
+		// Layout specifies whether sidebar or main content is being focused at this time.
+		const elementHandle = await this.page.waitForSelector( selectors.layout );
+		await elementHandle.waitForElementState( 'stable' );
+		const classAttributes = ( await elementHandle.getAttribute( 'class' ) ) as string;
+		const isSidebarOpen = classAttributes.includes( 'focus-sidebar' );
+		assert.strictEqual( isSidebarOpen, true );
```
```suggestion
		const layoutElement = await this.page.waitForSelector( '.layout.focus-sidebar' );
		await layoutElement.waitForElementState( 'stable' );
```

Not sure if that stability check is necessary though as the `focus-*` class seems to be added after the animation finished.

worldomonation replied:
You're right, it was not necessary. I have tweaked your suggestion a little to reuse the defined selector for `.layout` portion, but otherwise your suggestion works without issue.

WunderBart commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -63,19 +74,26 @@ export class SidebarComponent extends BaseContainer {
+		// If mobile, sidebar is hidden by default and focus is on the content.
+		// The sidebar must be first brought into view.
+		if ( viewport_name === 'mobile' ) {
+			await this._openSidebar();
```
I think this check would make more sense inside the `openSidebar`, i.e. 

```js
async _openSidebarIfNecessary() {
	if ( viewportName !== 'mobile' ) {
		return;
	}
}
```

worldomonation replied:
I disagree; I think the method `openSidebar` should do only that, open the sidebar. 

My opinion is that methods and functions should be named in the definitive case instead of...whatever the opposite of that case is called 😛. So in this case, `openSidebar` is very clear in what it does - it opens the sidebar. It is called if the sidebar needs to be opened; whereas `openSidebarIfNecessary` leaves open room for interpretation on what 'necessary' means. 

The indefinite naming of the Selenium suite is something that has bugged me for some time, so this is my attempt at making everything more definite where possible. I'd be open to hearing your take on this though.

WunderBart replied:
`*IfNecessary` is a convention that is also used in Playwright, i.e. `ElementHandle.scrollToIfNeeded()` (a bit different wording there though). I understand the uncertainty issue though, so maybe at this point going with `openMobileSidebar` would be the best of both worlds?

worldomonation replied:
> *IfNecessary is a convention that is also used in Playwright, i.e. ElementHandle.scrollToIfNeeded()

To my knowledge that's the only time such naming convention is used in Playwright - and I fully understand that sometimes it's just not possible to be 100% consistent 😃 

> going with openMobileSidebar would be the best of both worlds?

Sure. I am not 100% on board with this name either because say a decision is made in the future that `tablet` viewport will also have hidden sidebar by default. The name then is no longer reflective of reality when the more catch-all name `openSidebar` would require no changes. 

With that said, that's a bridge we can cross when that time comes, so for now I can rename this method to reference the mobile viewport.

WunderBart replied:
>  I am not 100% on board with this name either because say a decision is made in the future that tablet viewport will also have hidden sidebar by default. The name then is no longer reflective of reality when the more catch-all name openSidebar would require no changes.

Good point! This way it definitely scales better and makes more sense from the user's perspective. Feel free to ignore my comment then 👍

worldomonation replied:
To be clear, I will stick with the renaming to `_openMobileSidebar` for now since the hypothetical situation I mentioned is unlikely to come to pass :)

worldomonation committed:
Address code review comments.
- change My Site to My Sites.
- simplify _openSidebar detection on if the sidebar is opened.

worldomonation committed:
For mobile, the subitem also must be specified.

worldomonation committed:
Additional waitForSelector to confirm the focus is back on the content.

worldomonation committed:
Rename `openSidebar` to reference mobile viewport.

worldomonation commented on test/e2e/specs/specs-playwright/wp-seo__preview-spec.js:
```diff
@@ -13,7 +13,7 @@ describe( DataHelper.createSuiteTitle( 'SEO Preview Page' ), function () {
 	it( 'Navigate to Tools > Marketing page', async function () {
 		const sidebarComponent = await SidebarComponent.Expect( this.page );
-		await sidebarComponent.gotoMenu( { item: 'Tools' } );
+		await sidebarComponent.gotoMenu( { item: 'Tools', subitem: 'Marketing' } );
```
The `Marketing` subitem is selected by default if one opens `Tools`, but this is good practice to enforce anyway since that behavior can change at any moment.

WunderBart reviewed (changes requested):
Just one more comment!

worldomonation committed:
Playwright: add mobile support to pages used by `seo_preview` spec. (#53964)

* Implement mobile support for MarketingPage and the `seo__preview` spec.

Parent branch: add/playwright-sidebar-mobile-support

* Rename selector from plural to singular.

worldomonation committed:
Remove redundant `_click` helper method from navbar.

worldomonation committed:
`this.page.click` not `this.click`.

WunderBart commented on packages/calypso-e2e/src/lib/pages/marketing-page.ts:
```diff
@@ -32,16 +33,7 @@ export class MarketingPage extends BaseContainer {
-	 */
-	async _postInit(): Promise< void > {
-		await this.page.waitForSelector( selectors.navtabsList );
+		super( page );
```
I think the constructor can be omitted if it's only calling `super` as it's inherited from the `BaseContainer` class.

WunderBart reviewed (approved):
Just one nitpick. Looks good! 🚢

worldomonation committed:
Remove unnecessary super constructor call.
